File: ts/kit/src/instructions/eata/undelegateEphemeralAtaPermission.ts
import { Address, Instruction, AccountMeta, AccountRole } from "@solana/kit";
import {
  EATA_PROGRAM_ID,
  MAGIC_CONTEXT_ID,
  MAGIC_PROGRAM_ID,
  PERMISSION_PROGRAM_ID,
} from "../../constants";
import {
  ephemeralAtaPdaFromOwnerAndMint,
  permissionPdaFromAccount,
} from "../../pda";

export interface UndelegateEphemeralAtaPermissionAccounts {
  payer: Address;
  owner: Address;
  mint: Address;
}

export async function createUndelegateEphemeralAtaPermissionInstruction(
  accounts: UndelegateEphemeralAtaPermissionAccounts,
): Promise<Instruction> {
  const ephemeralAta = await ephemeralAtaPdaFromOwnerAndMint(
    accounts.owner,
    accounts.mint,
  );
  const permission = await permissionPdaFromAccount(ephemeralAta);

  const accountsMeta: AccountMeta[] = [
    { address: accounts.payer, role: AccountRole.READONLY_SIGNER },
    { address: ephemeralAta, role: AccountRole.WRITABLE },
    { address: permission, role: AccountRole.WRITABLE },
    { address: PERMISSION_PROGRAM_ID, role: AccountRole.READONLY },
    { address: MAGIC_PROGRAM_ID, role: AccountRole.READONLY },
    { address: MAGIC_CONTEXT_ID, role: AccountRole.WRITABLE },
  ];

  const data = new Uint8Array([8]);

  return {
    accounts: accountsMeta,
    data,
    programAddress: EATA_PROGRAM_ID,
  };
}

File: rust/sdk/src/spl/instructions/undelegate_ephemeral_ata_permission.rs
use crate::{
    consts::{ESPL_TOKEN_PROGRAM_ID, MAGIC_CONTEXT_ID, MAGIC_PROGRAM_ID, PERMISSION_PROGRAM_ID},
    solana_compat::solana::{AccountMeta, Instruction, Pubkey},
    spl::EphemeralSplDiscriminator,
};

/// Undelegate an ephemeral ATA permission.
pub fn undelegate_ephemeral_ata_permission(
    payer: Pubkey,
    eata: Pubkey,
    permission: Pubkey,
) -> Instruction {
    Instruction {
        program_id: ESPL_TOKEN_PROGRAM_ID,
        accounts: vec![
            AccountMeta::new_readonly(payer, true),
            AccountMeta::new_readonly(eata, false),
            AccountMeta::new(permission, false),
            AccountMeta::new_readonly(PERMISSION_PROGRAM_ID, false),
            AccountMeta::new_readonly(MAGIC_PROGRAM_ID, false),
            AccountMeta::new(MAGIC_CONTEXT_ID, false),
        ],
        data: vec![EphemeralSplDiscriminator::UndelegateEphemeralAtaPermission as u8],
    }
}
